### PR TITLE
#5 NUnit 3 console should produce xml events for ITestEventListener which contain unique id in the scope of all test agents for NUnit 2 tests

### PR DESCRIPTION
**Pull request by NikolayPianikov**

### CODE REVIEW

File: src/extension/NUnit2FrameworkDriver.cs
Comment (CharliePoole) [inline src/extension/NUnit2FrameworkDriver.cs:80]: I'm not sure about this test. In fact, the driver requires no such thing. It's very happy with a null id. However, any runner that wants to use multiple instances of the driver will typically need to set it.

@rprouse What do you think is the right thing to do here? How will our decision affect future drivers that may be written? (Assuming we establish this as a general precedent)

**Review by CharliePoole — APPROVED** [review]
Seems fine to me but I'd like us to discuss the issue I raised before committing it. Specifically... should drivers interface specify something about what is needed as an id so that the engine can honor it? Or should we use a property for framework drivers to do the same thing? 

I'm OK to merge this but want us to create an issue that will eventually provide a more general solution.

File: src/extension/NUnit2FrameworkDriver.cs
Comment (rprouse) [reply src/extension/NUnit2FrameworkDriver.cs:80]: I am unsure too. If ID is required in one driver, then it should be required in all drivers, but that is technically an API change which I am against. 

That said, there is no description or explanation with this PR other than the title and there is no link to a related issue. That doesn't give me much context to work with. I can see that you did this work on the branch issue-5, but that isn't much to work with. Much better to put something like Fixes #5 in the description so we don't need to go hunting.

I am going to read through the related issues and try to get back to you soon.

Comment (CharliePoole) [reply src/extension/NUnit2FrameworkDriver.cs:80]: TestPackages always have an ID. It is set internally in the constructor and is always the string representation of an int.
DirectTestRunner, which is the only runner that deals with a driver, always sets the driver id to the package id. So both checks are actually redundant.

I have no problem with doing the checks as guard clauses and throwing ArgumentNullException or ArgumentException but we shouldn't throw an NUnitEngineException here. Any error is simply a programming error on our part.

**Comment by chris-smith-zocdoc** [comment]
From what I can tell, this is the same change that was merged into the [nunit-console repo](https://github.com/nunit/nunit-console/pull/38) back on August 24th. The initial import to [nunit-v2-framework-driver](https://github.com/nunit/nunit-v2-framework-driver/commit/addd878776efef37bdb3dfb0e7d226a2e34e4247) on August 31st doesn't appear to have included this change though.

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc As you know, the v2 driver is now a separate repo, which is why we created this separate PR to fix it. When a v2 test is run, the engine provides the driver with a unique prefix to be used for each assembly. The driver is responsible for actually making use of that id.

@NikolayPianikov There's an outstanding comment waiting for your reply, regarding what exception should be thrown. I can go either way however, planning on changing it later if we come up with a more general solution for drivers that require a specific format for the prefix. If we want to merge it, however, we still need a second approval. @rprouse what do you think?

**Comment by rprouse** [comment]
The code itself looks good. For the exception, since it is a programming error, I think that it should be a standard .NET exception, not sure which though as none are exact fits, but probably ArgumentException for the second (not an int) and either ArgumentException or NotSupportedException for the first? I lean towards AgumentException even though it is a property not an argument.

File: src/extension/NUnit2FrameworkDriver.cs
Comment (NikolayPianikov) [reply src/extension/NUnit2FrameworkDriver.cs:80]: @CharliePoole we already discussed this exception in my previous PR. And I thought that InvalidOperationException is preferable here, but after discussion I understood that NUnitEngineException it is more common way for NUnit (and also is more informative from my point of view)

**Comment by CharliePoole** [comment]
@NikolayPianikov I see your point... NUnitEngineException is already used in other places. I think we should keep it consistent by merging this and review at some later time.

I was a little bothered by the fact that the extension isn't actually part of the engine, but it does function as such, so maybe that's just a nitpick. 😄

**Comment by chris-smith-zocdoc** [comment]
I can confirm the fix is working for me @NikolayPianikov I built commit 338ab11e6d981cda511d3f6f29ed22b4a6802e2c and replaced the dlls locally, the output messages appear correct as does the teamcity interface. 

Will there be a separate release of nunit-v2-framework-driver on nuget? Does this require a version bump of the primary NUnit nuget package? 

Sorry for the dummy questions, I haven't been following the extension update and distribution discussions closely.

**Comment by CharliePoole** [comment]
@ChrisMaddock There is no plan for a release anytime soon. We would do a new release if this fixed a critical issue. As far as I can tell, no issue was ever filed, so we never triaged it and decided how high it's priority is.

I'm not sure what you mean by the "primary nunit package" but we generally bump versions right after we do a release. So, the our nuget packages are now at 3.6 and we are producing pre-releases with the -dev suffix.

**Comment by ChrisMaddock** [comment]
cc @chris-smith-zocdoc  - I don't think the above message was intended for me!

**Comment by CharliePoole** [comment]
My mistake! 😄

**Comment by chris-smith-zocdoc** [comment]
The original issue I filed is here https://github.com/nunit/teamcity-event-listener/issues/28

I meant the `NUnit.Console` package

The only one that really needs to be bumped is `NUnit.Extension.NUnitV2Driver` though, likely to 3.5.1

Not sure about the priority of the issue though, it is certainly a limited use case.

**Comment by CharliePoole** [comment]
I guess somebody should have moved the issue here so we would track it better. I'll label the PR so it's recorded fixed when we do release.

**Pull request merged**
